fe2-android/app/src/test/java/com/github/dedis/student20_pop/model/data/LAORepositoryTest.java

```diff
@@ -37,121 +43,201 @@
+    answerUnsubscribe.subscribe(testObserver);
+
+    // Check the correct DataSource is being used
     Mockito.verify(remoteDataSource, Mockito.times(1)).sendMessage(Mockito.any());
 
-    TestObserver<Answer> testObserverUnsubscribe = TestObserver.create();
+    // Check there is no answer before the response delay time
+    testScheduler.advanceTimeTo(RESPONSE_DELAY - 1, TimeUnit.MILLISECONDS);
+    testObserver.assertNotComplete();
 
-    answerUnsubscribe.subscribe(testObserverUnsubscribe);
+    // Check the unsubscribe result is ready at response delay time
+    // and verify the id of the result corresponds to the expected id
+    testScheduler.advanceTimeTo(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+    testObserver.assertComplete().assertNoErrors()
+        .assertValue(r -> r.getId() == REQUEST_ID);
+  }
+
+  @Test
+  public void testGetAllLaos() {
+    // Does not use LAOLocalDataSource
+
+    // Create the LAO and the subscriber to test from
+    Lao lao = new Lao(LAO_CHANNEL);
+    TestObserver<List<Lao>> subscriber = TestObserver.create();
 
-    testScheduler.advanceTimeBy(950, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertNotComplete();
+    // Subscribe to a LAO and wait for the request to finish
+    repository.sendSubscribe(LAO_CHANNEL);
+    testScheduler.advanceTimeBy(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
 
-    testScheduler.advanceTimeBy(100, TimeUnit.MILLISECONDS);
-    testObserverUnsubscribe.assertComplete().assertNoErrors()
-        .assertValue(r -> r.getId() == 42);
+    repository.getAllLaos().subscribe(subscriber);
+
+    // Check the LAO is present in the laoById map of LAORepository
+    List<Lao> repositoryLaoList = (List<Lao>) subscriber.getEvents().get(0).get(0);
+    assertEquals(repositoryLaoList.get(0).getChannel(), lao.getChannel());
   }
 
-  // Test the get methods: getLaoObservable and getAllLaos
   @Test
-  public void testGetLaos() {
+  public void testGetLaoObservable() {
     // Does not use LAOLocalDataSource
 
-    // Send subscribe to fill laoById and allLaoSubject in LAORepository
-    repository.sendSubscribe("/root/123");
-    TestObserver<Lao> testObserverLao = TestObserver.create();
-    TestObserver<List<Lao>> testObserverListLao = TestObserver.create();
+    // Subscribe to a LAO and wait for the request to finish
+    repository.sendSubscribe(LAO_CHANNEL);
+    testScheduler.advanceTimeBy(RESPONSE_DELAY, TimeUnit.MILLISECONDS);
+
+    // Set subscriber of the LAO list to test from
+    TestObserver<Lao> subscriber = TestObserver.create();
+    repository.getLaoObservable(LAO_CHANNEL).subscribe(subscriber);
 
-    testScheduler.advanceTimeBy(1500, TimeUnit.MILLISECONDS);
+    // Check the LAO is present in the allLaoSubject list of LAORepository
+    Lao repositoryLao = (Lao) subscriber.getEvents().get(0).get(0);
+    assertEquals(repositoryLao.getChannel(), LAO_CHANNEL);
+  }
+
+  @Test
+  public void testBroadcast() throws GeneralSecurityException {
```

Comment:
this test seems to be doing a lot of things, and it's actually quite hard to follow.
I can't quite put the finger on what's the issue, a few hypotheses:
- is the LaoRepository not easily testable? 
- are there too many dependencies and we end up having to mock things left and right? 
- would this test warrant a separate class with its own setup?
- would the testing code warrant some refactoring? In particular it seems the TestObserver could benefit from having a wrapper class, with a simple API along the lines of:
```
subscriber = new TestLaoSubscriber(repository, LAO_CHANNEL);
assertEquals(1, subscriber.GetEventsCount());
assertEquals(xxx, subscriber.GetEvent<T>(0).getXxx());
```